zonec.c

```diff
@@ -753,6 +753,426 @@ zparser_conv_nsec(region_type *region,
+		memcpy(buf, key + 3, key_len - 3);
+		buf[key_len - 3] = 0;
+		key_value = strtoul(buf, &endptr, 10);
+		if (endptr > buf	/* digits seen */
+		&& *endptr == 0		/* no non-digit chars after digits */
+		&&  key_value <= 65535)	/* no overflow */
+			return key_value;
+
+	} else switch (key_len) {
+	case sizeof("mandatory")-1:
+		if (!strncmp(key, "mandatory", sizeof("mandatory")-1))
+			return SVCB_KEY_MANDATORY;
+		if (!strncmp(key, "echconfig", sizeof("echconfig")-1))
+			return SVCB_KEY_ECHCONFIG;
+		break;
+
+	case sizeof("alpn")-1:
+		if (!strncmp(key, "alpn", sizeof("alpn")-1))
+			return SVCB_KEY_ALPN;
+		if (!strncmp(key, "port", sizeof("port")-1))
+			return SVCB_KEY_PORT;
+		break;
+
+	case sizeof("no-default-alpn")-1:
+		if (!strncmp( key  , "no-default-alpn"
+		            , sizeof("no-default-alpn")-1))
+			return SVCB_KEY_NO_DEFAULT_ALPN;
+		break;
+
+	case sizeof("ipv4hint")-1:
+		if (!strncmp(key, "ipv4hint", sizeof("ipv4hint")-1))
+			return SVCB_KEY_IPV4HINT;
+		if (!strncmp(key, "ipv6hint", sizeof("ipv6hint")-1))
+			return SVCB_KEY_IPV6HINT;
+		break;
+	default:
+		break;
+	}
+	if (key_len > sizeof(buf) - 1)
+		zc_error_prev_line("Unknown SvcParamKey");
+	else {
+		memcpy(buf, key, key_len);
+		buf[key_len] = 0;
+		zc_error_prev_line("Unknown SvcParamKey: %s", buf);
+	}
+	/* Although the returned value might be used by the caller,
+	 * the parser has erred, so the zone will not be loaded.
+	 */
+	return -1;
+}
+
+static uint16_t *
+zparser_conv_svcbparam_port_value(region_type *region, const char *val)
+{
+	unsigned long int port;
+	char *endptr;
+	uint16_t *r;
+
+	port = strtoul(val, &endptr, 10);
+	if (endptr > val	/* digits seen */
+	&& *endptr == 0		/* no non-digit chars after digits */
+	&&  port <= 65535) {	/* no overflow */
+
+		r = alloc_rdata(region, 3 * sizeof(uint16_t));
+		r[1] = htons(SVCB_KEY_PORT);
+		r[2] = htons(sizeof(uint16_t));
+		r[3] = htons(port);
+		return r;
+	}
+	zc_error_prev_line("Could not parse port SvcParamValue: \"%s\"", val);
+	return alloc_rdata_init(region, "", 0);
+}
+
+static uint16_t *
+zparser_conv_svcbparam_ipv4hint_value(region_type *region, const char *val)
+{
+	uint16_t *r;
+	int count;
+	char ip_str[INET_ADDRSTRLEN];
+	char *next_ip_str;
+	uint32_t *ip_wire_dst;
+	size_t i;
+
+	for (i = 0, count = 1; val[i]; i++) {
+		if (val[i] == ',')
+			count += 1;
+	}
+
+	/* count == number of comma's in val + 1 
+	 * so actually the number of IPv4 addresses in val
+	 */
+	r = alloc_rdata(region, 2 * sizeof(uint16_t) + IP4ADDRLEN * count);
+	r[1] = htons(SVCB_KEY_IPV4HINT);
+	r[2] = htons(IP4ADDRLEN * count);
+	ip_wire_dst = (void *)&r[3];
+
+	while (count) {
+		if (!(next_ip_str = strchr(val, ','))) {
+			if (inet_pton(AF_INET, val, ip_wire_dst) != 1)
+				break;
+
+			assert(count == 1);
+
+		} else if (next_ip_str - val >= (int)sizeof(ip_str))
+			break;
+
+		else {
+			memcpy(ip_str, val, next_ip_str - val);
+			ip_str[next_ip_str - val] = 0;
```

Comment:
This is a buffer overflow if next_ip_str - val == (int)sizeof(ip_str), right? Then that is equal to the size of the array, and this sets ip_str[one too far] = 0; and that can be solved by checking a bit more earlier.

Reply:
Oh wait no, that is exactly covered by the check if statement above, so should not happen. Sorry.